ldap/servers/slapd/back-ldbm/archive.c

```diff
@@ -490,3 +490,212 @@ ldbm_back_ldbm2archive(Slapi_PBlock *pb)
+                            dest_file, prerr, slapd_pr_strerror(prerr), return_value, bytes_to_write);
+                }
+                slapi_log_err(SLAPI_LOG_ERR, "archive_copyfile",
+                              "Failed to write (%s) error %d (%s) - real bytes %d, expected bytes: %d\n",
+                              dest_file, prerr, slapd_pr_strerror(prerr), return_value, bytes_to_write);
+
+                if (return_value > 0) {
+                    bytes_to_write -= return_value;
+                    ptr += return_value;
+                    slapi_log_err(SLAPI_LOG_NOTICE, "archive_copyfile",
+                                  "Retrying to write %d bytes\n", bytes_to_write);
+                    if (task) {
+                        slapi_task_log_notice(task,
+                                "archive_copyfile - Retrying to write %d bytes",
+                                bytes_to_write);
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+        if ((CPRETRY == i) || (return_value < 0)) {
+            return_value = -1;
+            break;
+        }
+    }
+
+error:
+    if (source_fd != -1) {
+        close(source_fd);
+    }
+    if (dest_fd != -1) {
+        close(dest_fd);
+    }
+    slapi_ch_free_string(&buffer);
+    slapi_ch_free_string(&dest_file);
+
+    return return_value;
+}
+
+static char *cert_files_600[] = {
+    "ca.crt", "Server-Cert.crt", "key4.db", "cert9.db", "pin.txt", "pwdfile.txt",
+    "noise.txt", NULL
+};
+static char *config_files_440[] = {"certmap.conf", "slapd-collations.conf", NULL};
+
+/* Archive nss files, 99user.ldif, and config files */
+int32_t
+ldbm_archive_config(char *bakdir, Slapi_Task *task)
+{
+    slapdFrontendConfig_t *config = getFrontendConfig();
+    char *schema_dir = config->schemadir;
+    char *cert_dir = config->certdir;
+    char *config_dir = config->configdir;
+    char *backup_config_dir = slapi_ch_smprintf("%s/config_files", bakdir);
+    char *dse_file = slapi_ch_smprintf("%s/dse.ldif", config_dir);
+    char *schema_file = slapi_ch_smprintf("%s/99user.ldif", schema_dir);
+    int32_t rc = -1;
+
+    /* Create config_files directory */
+    if (mkdir(backup_config_dir, 0700) != 0) {
+        slapi_log_err(SLAPI_LOG_ERR, "ldbm_archive_config",
+                "Failed to create 'config_files' directory in the archive location. Error %d\n",
+                errno);
+        if (task) {
+            slapi_task_log_notice(task,
+                    "Failed to create 'config_files' directory in the archive location. Error %d",
+                    errno);
+        }
+        slapi_ch_free_string(&backup_config_dir);
+        slapi_ch_free_string(&dse_file);
+        slapi_ch_free_string(&schema_file);
+        return rc;
+    }
+
+    /* Config dir - dse.ldif */
+    if (archive_copyfile(dse_file, backup_config_dir, "dse.ldif", 0600, task) != 0) {
```

Comment:
I am a bit worried about the on line backup and the dse.copy:
 IMHO we should insure that the dse.ldif file is not written while the copy is taking place.

Reply:
my concern is that during backup all backend instances are marked as busy but unless mistaken 
there may still be running operations targeting cn=config or cn=schema and these operation may rewrite the files with a risk of having empty or half written file when they get copied.

Reply:
You can checksum the file before you start the copy, do the copy and the checksum the destination. If they are the same, it's a valid copy, if not it means it's been altered or truncated. Then if that occurs, you re-attempt the copy. You can use crc32c pretty easily for this or something else, but it will resolve the concern.

Reply:
No the checksum idea does not resolve the concern: If another  thread is writing the files it starts by emptying it then the checksum, copy, verify checksum will be very fast. They may even be completed before the thread that is writing the file has written the first block.  IMHO we should rather use locks  (I think there are already existing locks that protect against dse or schema changes)

Reply:
If there are locks then we should use them yes.

Reply:
> No the checksum idea does not resolve the concern: If another thread is writing the files it starts by emptying it then the checksum, copy, verify checksum will be very fast. They may even be completed before the thread that is writing the file has written the first block. IMHO we should rather use locks (I think there are already existing locks that protect against dse or schema changes)

We need a new global lock for this.  The current lock implementation is per dse update session.  It will still be easy to implement and cover schema updates as well.

Reply:
Do you suggest to use nsslapd-global-backend-lock ? I thought it also covers schema updates.

Reply:
> Do you suggest to use nsslapd-global-backend-lock ? I thought it also covers schema updates.

I was implementing a new lock in dse.c, but I'm not sure the global backend lock covers cn=config?  I'll have to check that, but I don't think it does.

Reply:
Fixed it with a new lock